refactor(linter/plugins): ensure debug assertions always shaken out of release build

overlookmotel opened the pull request:
The `assert*` functions are no-ops in release builds, and minifier removes them as dead code.

However, there's an annoyance - we have to ensure that they're only used in files which end up in the `plugins.js` chunk. If we use those assertions elsewhere (e.g. in `cli.ts`), TSDown creates a shared chunk containing them, and then minifier can't see that they can be removed, and leaves all the assertions in the output.

Fix this problem by adding a plugin to TSDown for the release build which replaces `import`s from `asserts.ts` with inlined empty functions.

```ts
// Original code
import { assertIs, assertIsNonNull } from '../utils/asserts.ts';
```

```ts
// After transform
function assertIs() {}
function assertIsNonNull() {}
```

This allows us to use the assertion functions anywhere, and minifier can always remove them as dead code.

camc314 commented:
### Merge activity

* **Nov 21, 2:54 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Nov 21, 2:56 PM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
refactor(linter/plugins): ensure debug assertions always shaken out of release build (#15946)

The `assert*` functions are no-ops in release builds, and minifier removes them as dead code.

However, there's an annoyance - we have to ensure that they're only used in files which end up in the `plugins.js` chunk. If we use those assertions elsewhere (e.g. in `cli.ts`), TSDown creates a shared chunk containing them, and then minifier can't see that they can be removed, and leaves all the assertions in the output.

Fix this problem by adding a plugin to TSDown for the release build which replaces `import`s from `asserts.ts` with inlined empty functions.

```ts
// Original code
import { assertIs, assertIsNonNull } from '../utils/asserts.ts';
```

```ts
// After transform
function assertIs() {}
function assertIsNonNull() {}
```

This allows us to use the assertion functions anywhere, and minifier can always remove them as dead code.
